tests/ci/ci_cache.py

```diff
@@ -723,30 +724,30 @@ def filter_out_not_affected_jobs(self):
                     job_config=reference_config,
                 ):
                     remove_from_to_do.append(job_name)
+                    has_test_jobs_to_skip = True
                 else:
                     required_builds += (
                         job_config.required_builds if job_config.required_builds else []
                     )
 
-        has_builds_to_do = False
         for job_name, job_config in self.jobs_to_do.items():
             if CI.is_build_job(job_name):
                 if job_name not in required_builds:
                     remove_from_to_do.append(job_name)
-                else:
-                    has_builds_to_do = True
 
-        if not has_builds_to_do:
+        if not required_builds:
             remove_from_to_do.append(CI.JobNames.BUILD_CHECK)
 
-        for job in remove_from_to_do:
-            print(f"Filter job [{job}] - not affected by the change")
-            if job in self.jobs_to_do:
-                del self.jobs_to_do[job]
-            if job in self.jobs_to_wait:
-                del self.jobs_to_wait[job]
-            if job in self.jobs_to_skip:
-                self.jobs_to_skip.remove(job)
+        if has_test_jobs_to_skip:
```

Comment:
it's a bit tricky. `remove_from_to_do` also has builds (see 733-736) that are not required by any test job (all builds from Builds_2 are not required). If _has_test_jobs_to_skip_ is False it means we don't need to apply a filter and drop anything, otherwise we'll skip all builds from Builds_2. I'll try to reorganise this code to make it more transparent

Reply:
ah I see now we add to `remove_from_to_do` after the flag was set, everything is okay

Reply:
I changed code a bit to make it more clear, also added this function to a unit test